machines: Make readonly and shareable disk options configurable

skobyda opened the pull request:
Allow multiple running VMs share a single disk.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1684304
Closes #12314

skobyda commented:
![Screenshot from 2019-07-12 17-10-20](https://user-images.githubusercontent.com/42733240/61143014-67e9b380-a4d1-11e9-8107-6ef274d94e9b.png)

### This functionality takes effect only after shutting down the VM.
![Screenshot from 2019-07-12 17-10-38](https://user-images.githubusercontent.com/42733240/61143029-6ddf9480-a4d1-11e9-878e-2268a59d5b26.png)

### Disks which are not persistent (3rd disk in row) do not have configurable options:
![Screenshot from 2019-07-12 18-05-33](https://user-images.githubusercontent.com/42733240/61143095-96678e80-a4d1-11e9-891d-b8f8f8f562f7.png)

KKoukiou commented:
This affects the design of the disk row as well, @garrett are you good with this change?

garrett commented:
1. We shouldn't have checkboxes without obvious labels.
2. When would someone want read-only disks? Would they ordinarily toggle them? Is it changeable on the fly or only after a reboot? If it's possible to change a disk from RO to RW (or vice versa) when running, would things prevent that? Are errors handled, if so?
3. What does "Sharable" mean?
4. The values under "source" are oddly aligned.
5. There's too much space on the [-] buttons. (I can look into and fix this. It's probably a bit of CSS.)

skobyda commented:
> 1. We shouldn't have checkboxes without obvious labels.

For readonly, what about: Forbid/allow modification of disk.
> 2. When would someone want read-only disks?

Usually when user wants multiple VMs to use the same disk. To prevent write conflicts, you set this disk as readonly for some VMs, so they loose write access. Or disks which you want to act as "Cdrom" or other Readonly disks.
> Would they ordinarily toggle them?

Again, mainly for specific use cases such as multiple VMs share disk, or to simulate readonly disk type.
> Is it changeable on the fly or only after a reboot? If it's possible to change a disk from RO to RW (or vice versa) when running, would things prevent that? Are errors handled, if so?

No, this can be only updated after a reboot. You can modify readonly/shareable options only for permanent VM XML, which will be loaded with all the new modifications after VM reboots. Good idea, I forgot to take into account error handling (altought I can't think of any error apart from some undocumented libvirt/qemu internal error)
> 3. What does "Sharable" mean?

Again, used when you want to share a disk with multiple VMs, but this case you want them to have a write access. From what I can read in documentation, it will turn off I/O caching, locking of the disk which will make disk available for modification by multiple VMs. It's mainly used for cluster-aware guests.  
> 4. The values under "source" are oddly aligned.
> 5. There's too much space on the [-] buttons. (I can look into and fix this. It's probably a bit of CSS.)
 
Both of these things are not affected by this PR.

skobyda commented on pkg/machines/libvirt-common.js:
```diff
@@ -884,6 +884,64 @@ export function unknownConnectionName(action, libvirtServiceName) {
+export function updateDiskReadOnly(domXml, diskTarget) {
+    const domainElem = getDomainElem(domXml);
+    if (!domainElem)
+        throw new Error("updateDiskReadOnly: domXML has no domain element");
```
To be honest I put it there only because I have noticed it in other updateXML functions, so I thought there is reason for that. But now when I read about it a bit, libvirt would not allow this scenario, and since we pull all the XML for updateXML functions from libvirt, I guess it's safe not to have it.

skobyda commented on pkg/machines/components/vmDisksTab.jsx:
```diff
@@ -54,14 +58,40 @@ const VmDiskCell = ({ value, id }) => {
     );
 };
 
+const readonlyChanged = (vm, diskTarget) => {
```
Sure, talking about merging, the updateDiskReadonly and updateDiskShareable (or toggle...) could be merged too, since only difference is in different XML element name. What do you think?

skobyda commented:
@garrett After our IRC discussion to pull readonly and shareable configuration into separate disk Edit dialog, here is the first take of how it could look:
![Screenshot from 2019-07-17 12-52-46](https://user-images.githubusercontent.com/42733240/61370064-cfb54b00-a891-11e9-96dc-355d2f20fd13.png)

Alternatively checkboxes can be replaced by dropdown with options (Read-only / Read-write) and (Shareable / Non-shareable).
@KKoukiou What do you think about introducing Edit Disk dialog?

garrett commented:
@skobyda: Cool! Thanks for working on splitting the state changing into a dialog.

1. Are we going to have additional ways to edit the disk in the future?
2. How are we editing the disk? "Edit disk" might not be specific enough. (On the other hand, we want to avoid saying things like "mount" in the UI.) "Edit disk settings" or "Edit disk state" perhaps? :man_shrugging:
3. Which disk is this? There's no indication what is being edited. (We should add the name to the title or above the checkboxes.)
4. Can you share read-write disks?
5. "Readonly" should be "Read only" or "Read-only".
6. Will this be instant-apply, or will a reboot have to happen for this to take effect? We should say if a reboot is needed. :wink:
7. Is there an indication that a reboot is needed for this to take effect after the disk mount settings have been applied?

skobyda commented:
> 1. Are we going to have additional ways to edit the disk in the future?

Yes, you can edit stuff like Bus, storage format and several performance options.
> 2. How are we editing the disk? "Edit disk" might not be specific enough. (On the other hand, we want to avoid saying things like "mount" in the UI.) "Edit disk settings" or "Edit disk state" perhaps? man_shrugging
> 3. Which disk is this? There's no indication what is being edited. (We should add the name to the title or above the checkboxes.)

Disk is uniquely specified by it's path. So dialog name could be something like "/var/lib/libvirt/images/fedora30.qcow2 Disk Settings".
> 4. Can you share read-write disks?

Yes. 
> 5. "Readonly" should be "Read only" or "Read-only".
> 6. Will this be instant-apply, or will a reboot have to happen for this to take effect? We should say if a reboot is needed. wink

It wil happen only after reboot. I will put warning at the dialog footer similar to Vcpu or Network Interface Edit dialogs. 
> 7. Is there an indication that a reboot is needed for this to take effect after the disk mount settings have been applied?
Yes, the warning will show in disks table next to property which value will change after reboot.

garrett commented:
What if the dialog would be something like this:

Edit disk state | ×
--: | --
Name | Disk-123
Size | 6GB
State | ● Read & write<br>○ Read-only<br>○ Read-only and shared

Name and size are not editable (at least in this iteration). They're in the dialog for reference, to make sure someone knows what they're editing.

State is a group of 3 radio buttons.

If "read-only and shared" is selected, then perhaps there should be a warning message about how this makes the disk read-only until it is removed from other VMs. It would also be good to know which VMs are using the disk (from _somewhere_, within Cockpit).

The × is centered because this is a limitation of doing mockups in tables in Markdown on GitHub. Same for other alignment issues. :wink:

What do you think?

marusak commented:
Needs to rebase to master since #12367 changed tests names

skobyda commented:
@garrett Current state of dialog:
![Screenshot from 2019-07-23 13-02-06](https://user-images.githubusercontent.com/42733240/61707168-1bf90300-ad4a-11e9-8d77-3e5b87625bb1.png)

Also as you requested I moved the action buttons to the right, replaced words yes/no -> Read-Only/Read-Write or Shareable/Not Shareable. And also changed the [-] button into [Detach], so it fits more nicely together with [Edit] Button.

**Now**:
![Screenshot from 2019-07-23 13-02-58](https://user-images.githubusercontent.com/42733240/61707376-a2ade000-ad4a-11e9-9fd1-b21cc0e08905.png)

**Before**:
![Screenshot from 2019-07-23 13-04-17](https://user-images.githubusercontent.com/42733240/61707399-af323880-ad4a-11e9-96b8-968951987820.png)

garrett commented:
The dialog:
- Lacks a name, ID, path, or whatever else (like we talked about in IRC earlier) to show which disk is being edited
- Has confusing text
- Has text that expresses one thing (that sharing is about writing) where the UI widgets express another (sharing is independent of writing)

Edit disk attributes |   |  
--: | -- | --
Name | boot-scratch/test-disk
Access | ● Read and write | ○ Read-only
Share | 🗹 Allow multiple Vms to access this disk |  

Is detaching like unmounting? How do you remove the disk?

garrett commented:
One issue that might arise with using the word "share": It could imply filesharing like NFS, Samba, etc.

skobyda commented:
> The dialog:
> 
> * Lacks a name, ID, path, or whatever else (like we talked about in IRC earlier) to show which disk is being edited

Yes, working on it right now

> * Has confusing text

> * Has text that expresses one thing (that sharing is about writing) where the UI widgets express another (sharing is independent of writing)
> 
> Perhaps it should be more like this:
> 
> Edit disk attributes	 
> Name	boot-scratch/test-disk
> Access	● Read and write   ○ Read-only
> Share	🗹 Allow multiple Vms to access this disk

This mockup looks good. I would only change sentence "Allow multiple Vms to access this disk", since even without shareable option, multiple Vms can access disk in read-only mode. with shareable option, multiple VMs can write into this this. So perhaps: "Allow multiple Vms to write to this disk"

> "Is detaching like unmounting? How do you remove the disk?" This is what our users will think.

Detach/Attach is terminology used mainly with disks, while mount/umnount with filesystems. Altought both are similar.

> In a UI, you need to use language that mirrors itself. We have "Add disk," so it should be "Remove", not "Detach". (We'd have to have "Attach disk" to have "Detatch". And even then, Add/Remove is better & friendlier than Attach/Detatch.)
> 
> Please look over [PatternFly's style guide](https://www.patternfly.org/v4/design-guidelines/content/grammar-and-terminology).

If we use Patternflys guide, then Remove would be best option, since Detaching is unassigning Disk from VM, but not deleting it yet.

skobyda commented:
So, here's updated version. Let me know if I forgot about anything or something is unclear:

Disks defined as file by file path:
![Screenshot from 2019-07-23 18-08-02](https://user-images.githubusercontent.com/42733240/61728125-e584ad80-ad74-11e9-9b1f-412f9228b6e1.png)

Disks defined by storage pool/name:
![Screenshot from 2019-07-23 18-08-11](https://user-images.githubusercontent.com/42733240/61728131-e74e7100-ad74-11e9-93a6-fb09a2895952.png)

Some disks do not have any identifier or anything a bit unique to use as name, so for this cases fallback is to use device type:
![Screenshot from 2019-07-23 18-08-23](https://user-images.githubusercontent.com/42733240/61728139-e9b0cb00-ad74-11e9-9162-3f294b94c0d6.png)

KKoukiou commented:
> > The dialog:
> > 
> > * Lacks a name, ID, path, or whatever else (like we talked about in IRC earlier) to show which disk is being edited

Shouldn't it be in the Modal title?

> 
> Yes, working on it right now
> 
> > * Has confusing text
> 
> > * Has text that expresses one thing (that sharing is about writing) where the UI widgets express another (sharing is independent of writing)
> > 
> > Perhaps it should be more like this:
> > Edit disk attributes	 
> > Name	boot-scratch/test-disk

I think 'Name' label is misleading since disks don't have a name. They have path, pool. volume, and other stuff.
Since we devices the pool/volume or path should be the identifier I suggest this should stay on the modal title, and inside the dialog the entries presented should correspond to libvirt XML elements, not made up ones.

> > Access	● Read and write   ○ Read-only
> > Share	🗹 Allow multiple Vms to access this disk
> 
> This mockup looks good. I'm only not sure about sentece "Allow multiple Vms to access this disk", since even without shareable option, multiple Vms can access disk in read-only mode. with shareable option, multiple VMs can write into this this. So perhaps: "Allow multiple Vms to write to this disk"?
>

Pasting the text from the documentation for the shareable attribute:
```
Critical for the disk to be safely shared between guests, this ensures that the SELinux labelling will be appropriate for multiple guests to access the disk and that all I/O caching is disabled on the host. 
```
 
> > "Is detaching like unmounting? How do you remove the disk?" This is what our users will think.
> 
> Detach/Attach is terminology used mainly with disks, while mount/umnount with filesystems. Altought both are similar. Detach/Attach is also widely used with disks in virtualization, but it means same as Remove/Add from(into) VM.
> 
> > In a UI, you need to use language that mirrors itself. We have "Add disk," so it should be "Remove", not "Detach". (We'd have to have "Attach disk" to have "Detatch". And even then, Add/Remove is better & friendlier than Attach/Detatch.)
> > Please look over [PatternFly's style guide](https://www.patternfly.org/v4/design-guidelines/content/grammar-and-terminology).
> 
> If we use Patternflys guide, then Remove would be best option, since Detaching is unassigning Disk from VM, but not deleting it yet.


I wouldn't replace 'Detach' with 'Remove'.
I prefer replacing 'Add' with 'Attach'.
'Attach'/'Detach' are terms commonly used when referring to the process of adding/removing devices to VMs, even in RHEL documentation.

In addition Remove is not clear if the disk gets deleted as well. Remove from the guest or Remove from the system?

KKoukiou commented:
> So, here's updated version. Let me know if I forgot about anything or something is unclear:
> 
> Edit: Ups, didn't notice there is not `<hr />` between **Name** and **Access**. Fixed that.
> 
> Disks defined as file by file path:
> ![Screenshot from 2019-07-23 18-08-02](https://user-images.githubusercontent.com/42733240/61728125-e584ad80-ad74-11e9-9b1f-412f9228b6e1.png)
> 
> Disks defined by storage pool/name:
> ![Screenshot from 2019-07-23 18-08-11](https://user-images.githubusercontent.com/42733240/61728131-e74e7100-ad74-11e9-93a6-fb09a2895952.png)
> 
> Some disks do not have any identifier or anything a bit unique to use as name, so for this cases fallback is to use device type:
> ![Screenshot from 2019-07-23 18-08-23](https://user-images.githubusercontent.com/42733240/61728139-e9b0cb00-ad74-11e9-9162-3f294b94c0d6.png)

I find it a bad idea to have the 'Name' label like this.
The identifier should be used as is in the modal title, but inside the dialog body, only real XML attributes should be presented. (Sorry I caught this after 2 review rounds)
Name = cdrom is a bit nonsense, really. Same for Name: poolName/volName

Just keep them as in the XML, like:

Type network 
Device cdrom

or Type file
    Pool myPool
    Volume myVolume

Of course let @garrett confirm this.

skobyda commented:
> I think 'Name' label is misleading since disks don't have a name. They have path, pool. volume, and other stuff.
> Since we devices the pool/volume or path should be the identifier I suggest this should stay on the modal title, and inside the dialog the entries presented should correspond to libvirt XML elements, not made up ones.
> 

There are 4 types of labels we can use instead of name. For:
- _file/bock/dir_ disks identifier is a **path**
- _network_ disks identifier is a mostly **url**
- _volume_ disks identifier is a **pool/volume**
- and as a fallback, we can use identifier as a combination **type** (e.g file) and ### device (e.g. cdrom)


> ```
> Critical for the disk to be safely shared between guests
> ```

I like this sentence. Perhaps we can use it as next to Share checkbox? @garrett

garrett commented:
- I didn't really mean to use the label "name" unless it's actually the label that we're supposed to use. I meant that we need to show some information to identify the disk.
- If there's a reliably short identifier, it's enough to have it in the titlebar.
- If there isn't a reliably short identifier, that's when we need to have the information in the dialog body too.
- Please remove the `<hr>`s. They're not supposed to be used in a `ct-form` unless there are unrelated fields next to each other.
- Perhaps we should use "Sharing" instead of "Share" (looking at it here, "Share" looks weird).
- We're supposed to use "Sentence case" in PF4 design guidelines, not "Title Case" — however, as we use "Title Case" for so many things all over the place, and this would require changing everywhere, it's fine to use "Title Case" (where PF3 dictates, like in titles) for now.
- These aren't really "settings". They're more "attributes" of the disk.
- It's redundant to have "disk" in the title.
- The title should probably be something like "Edit $DiskNameHere attributes" — but "$DiskNameHere attributes" could also work.

skobyda commented:
So if I inderstand it correctly, if we can use some short identifier in title, like (**filename** for block/dir/file disk or **volumename** for pool disk), then we should not add any extra attribute. But if there is no senslible way how to get short identifier (**url** for network disk or **type + device** as fallback), then is should be shown only in dialog body and title should just contain "Edit attributes" ?

garrett commented:
@skobyda: The idea is that there should be some way for a person to know what they're actually editing within the UI itself. It's too easy to click on the wrong item and edit the wrong thing if the dialog doesn't let you know that.

garrett commented:
Reminders for Tuesday, when we'll talk about this again:

- It seems like sharable is only for multiple read-write systems — @KKoukiou is going to check on this and follow up here.
- Cockpit should consider all VMs and their disks and preventatively try to fix any sharable issues
   - libvirt throws an error; we should handle it appropriately and try to fix
   - First pass should only be the options dialog, but we should spec out what would happen in the grand scheme of things
   - A follow-up PR would add a friendlier flow for toggling sharable when it is needed
   - This is a little tricky, as it looks like we can change the value on the fly but it won't take effect until affected VMs are rebooted
- Perhaps we should re-visit the 3 options (if sharable is only for r/w)
- We should reword "sharable" and make it clear what this is doing, why it is needed, and what the drawbacks are

KKoukiou commented:
Update from IRC conversation:
```
kkoukiou: since the vrit driver requires either readonly or shareable, not both - it doesn't make sense for a disk on a single guest to be readonly-shared and readwrite-shared at the same time
kkoukiou: what you're talking about is slightly different though as you have multiple guests in play
kkoukiou: you are correct though, if one guest needs it shared-write, then all guests must use it shared-write
```
So, we should only allow readonly, readwrite, and readwrite-shared.
Another point that we have to make clear in the UI, is that these options are only from the host OS POV. The guest can actually mount a disk that in the XML is readonly as readwrite.
IOW we should communicate it to the user that when mounting the disk  they should respect the choices they made in the UI, because nothing will prevent them from not doing that.

skobyda commented:
> ```
> kkoukiou: you are correct though, if one guest needs it shared-write, then all guests must use it shared-write
> ```

So in this case, initial mockup submitted by @garrett contaiing 3 options is best option:

> ○ Read-only
> ● Read-write
> ○ Read-write and shared

garrett commented:
> The guest can actually mount a disk that in the XML is readonly as readwrite.

I'm wondering about rewording the 3 options as a result. The read-write isn't the important part here, it's shared.

> ○ Read-only
> ● Read-write
> ○ Shared between multiple VMs 🛈

…And then the little 🛈 could have additional information about the ability to be read-write or other important details.

What should happen when a disk is marked as shared in multiple VMs and is then changed to not-shared in one VM? It can't be both shared and not shared.

Is this actually editing the details of the disks across all VMs? Is it just for the disk inside the current VM?

skobyda commented:
> Is this actually editing the details of the disks across all VMs? Is it just for the disk inside the current VM? (We talked about a follow-up PR to handle various cases a bit better... but I want to know what happens meanwhile.)

This is just for disk inside the current VM. If you do conflicting configuration (e.g. shared for one, and not shared for other), then standardly libvirt would return error when you try to boot VM with conflicting disk configuration, and as a result VM won't boot. 

> What should happen when a disk is marked as shared in multiple VMs and is then changed to not-shared in one VM? It can't be both shared and not shared.

What should happen. Well, if standart libvirt error is not good for us, we can predict conficting disk configurations. I can think of alternatives:
- If VM1 and VM2 have conflicting disk configuration, and VM1 is already running, we can disable "Run" button for VM2 with tooltip which says something like "VM1 disk configuration 'not-shared' is in conflict with VM2 configuration 'shared'".
- If users opens VM1's "Edit disk attributes" dialog, and changes some configuration into something which we can predict would be conflicting, we can show warning which says something like: "This configuration is in conflict with VM2 disk configuration. If you do not change VM2 configuration, then both vms won't be able to run at the same time".
- Or we can implement both above mentioned options.

garrett commented:
@skobyda: We talked about the conflict resolution already; I'm wondering how much of that should go into this PR.

In general: We shouldn't just show an error or a warning; we should provide a way to solve the problem (or just solve it up-front automatically or even prevent it, if possible).

skobyda commented:
> @skobyda: We talked about the conflict resolution already; I'm wondering how much of that should go into this PR.

Oh, I thought that you wanted some temporary solution until follow-up PR.

> In general: We shouldn't just show an error or a warning; we should provide a way to solve the problem (or just solve it up-front automatically or even prevent it, if possible).

- **To solve the problem**: From our call, I like the most the solution where if user tries to run a VM which has conflict, then a window would popup. From this window user could change the disks configuration for all the VMs which is conflicting. But this really is a thing for follow-up PR.
- **To solve it up-front automatically**: would mean to change disk configuration automatically without user knowing, which when I think about it I don't whink would be good idea.
- **To prevent it**: The problem only happens once you try to run conflicting VM. So disabling "Run" button would be possible, as I have mentioned in my previous post.

The prevention option could be combined with follow-up PR.
In **this** PR, as a prevention, we could just disable Run button for conflicting VM, with some explanation.
In **follow-up** PR, we could add next to that disabled Run button an option to solve the problem (e.g. popup window where user could change the disks configuration for all the vms which is conflicting)

So in this PR we will have only prevention of the problem. And after follow-up PR, we will have prevention of the problem & option to solve of the problem for user.

garrett commented:
> some temporary solution until follow-up PR

Yeah, it'd be nice. :+1:

> So disabling "Run" button

I meant prevention of getting into the problem in the first place. Of course, this would only work if Cockpit is used exclusively for managing VMs, so yeah, having a fallback error state is important.

However, disabling the run button would mean that someone can't run a container and won't know why. There would have to be some way to show that there's an error state, explain it enough, and provide a way to fix things.

But yeah, this is probably still for that future PR.

> So in this PR we will have only prevention of the problem. And after follow-up PR, we will have prevention of the problem & option to solve of the problem for user.

Yep; great.

skobyda commented:
> > some temporary solution until follow-up PR
> 
> Yeah, it'd be nice. 
> 
> > So disabling "Run" button
> 
> I meant prevention of getting into the problem in the first place. Of course, this would only work if Cockpit is used exclusively for managing VMs, so yeah, having a fallback error state is important.
>
> However, disabling the run button would mean that someone can't run a container and won't know why. There would have to be some way to show that there's an error state, explain it enough, and provide a way to fix things.

Of course there would be and explanation why user cannot run a VM.
> 
> But yeah, this is probably still for that future PR.

There is no way how to prevent it earlier. User can safely has conflicting configuration of disks on different VMs as long as they are **not running simultaneously**, so preventing a conflicting configuration when editing disk is not a good idea as there are cases when it's safe (e.g. VMs are not running at the same time).
So I can think of temporary options which can be done in this PR (if you want to):
1. If user chooses conflicting configuration in Disk Edit Dialog, show some warning which says there might be a problem if you run this machine together with ... machine because it's conflicting.
2. Disable "Run" button and show explanation why cannot

skobyda commented:
Alright, let's revive this PR!

So to sum it up, we will introduce disk setting **Access** with three options: `Read-only`, `Read-Write` `Read-Write and shared`, which represents various combinations of libvirt's readonly and shareable settings.
![Screenshot from 2019-09-05 15-15-38](https://user-images.githubusercontent.com/42733240/64345482-bc507180-cff0-11e9-89fb-93ed17ece667.png)


The problem was that when user added as disk a volume which was already used as a disk by some other vm, they have 2 options how to use those 2 VMs using same disk:
1. They would never run all VMs simultaneously. This is a perfectly fine use case for when user wants to have basically same machine (same disk) with just two different configuration (two differently configured VMs), and run either VM depending to their need. This option doesn't cause any problem for us/
2. They would want to run all VMs using that disk simultaneously. In that case user needs to change disk configuration setting on every VM to "Read-Write and shared" separately, which is a nuisance. So as a solution, I recommend giving option during adding of disk. If we detect that a disk which user wants to add to VM is already used by another VM(s), we offer to change disk configuration on all other VMs to "Read-Write and shared"
![Screenshot from 2019-09-05 15-21-06](https://user-images.githubusercontent.com/42733240/64345552-de49f400-cff0-11e9-8d2b-9179396b0c99.png)
![Screenshot from 2019-09-05 15-21-25](https://user-images.githubusercontent.com/42733240/64345559-e144e480-cff0-11e9-84b1-0a9c9d687d30.png)

garrett commented:
## Disk attributes dialog

- Access attributes should be a set of vertical radio buttons
- Storage volume name doesn't need to be monospace

![vm-disk-sharing-edit-attributes](https://user-images.githubusercontent.com/10246/64360525-f2025400-d00a-11e9-8e17-d7ea356f8e01.png)

## Add disk dialog

- Having a confirmation prompt in the dialog is a bit odd.
   - When we talked about this last, the dialog would automatically change things so everything works. It should mention this in the dialog itself, however.
- I don't have a mockup for this yet. (It's a work in progress and isn't ready to share. I ran out of time to finish it this week. Sorry. :cry:)

skobyda commented:
>   * When we talked about this last, the dialog would automatically change things so everything works. It should mention this in the dialog itself, however.

But I believe that we should offer option to user so they can choose whetever they want us to change disks to shareable automatically or not.

Reason is that user might not always want us to change shareable option on disks automatically. as shareable option brings some downsides. E.G. I/O caching turns off, which slowns down the disk (and after googling around, I see that a lot of people care about their virtual disks performance) and it would disable ability of creating a snapshots of said disks (which is a thing we want to implement in future).

And as I described in my previous comment, user can have use case where they might not run VMs simultaneously and thus they do not need shareable option. Therefore if we turn that option automatically without user's confirmation of at least letting them know so they can change it back we might unnecessary bring all these downsides upon them.

So if user wants to run VMs sharing disk simultaneously, it's good and useful we change shareable options automatically.
If user doesn't want to run VMs simultaneously, it might be harmful to change these options of disks.
That's why I suggested that confirmation prompt.

This is a option which will probably need a lot of explanation in UI :(.

garrett commented:
> But I believe that we should offer that option or at least inform user what is going to happen.

Yes, we should still inform people what would happen.

> […] might not always want us to change shareable option on disks automatically. as shareable option brings some downsides

Not turning it on the flag, but turning on another VM with the same disk would _definitely_ cause disk / data corruption, which is completely irresponsible on our part.

What you're suggesting is to ask our users to keep track of state in their heads over the course of a long time — as long as their VMs exist... and then for them to turn one off before turning another on. That's also unacceptable.

One of the major points of Cockpit is to make things easy and to also prevent people from "shooting themselves in the foot". (That is, we shouldn't cause someone to accidentally do something devastating.)

skobyda commented:
Alright I see your point. If we give big enough warning which will tell user what's gonna happen, then I'm fine with that.
Something like: "This volume is already used by: .... Attaching it will make this disk shareable for every VM" ?

skobyda commented:
A bit like:
![Screenshot from 2019-09-11 13-38-37](https://user-images.githubusercontent.com/42733240/64694236-86503900-d499-11e9-9c1f-1ca66172a801.png)

garrett commented:
> Also let's not forget that this whole "sharing disks" stuff only works for raw disks (not for qcow...)

Then don't show the shared option, please.

skobyda commented:
Updated. UI now doesn't show sharing option for non-raw disks.
![Screenshot from 2019-09-16 12-41-23](https://user-images.githubusercontent.com/42733240/64951991-72cd1580-d87f-11e9-9e56-c96e1d1be46c.png)

skobyda commented:
Tests seem to be green. It also seems there are no more design objections.
@KKoukiou once you have some free time, may I ask for code review?

KKoukiou commented on pkg/machines/helpers.js:
```diff
@@ -676,3 +676,77 @@ export function getDefaultVolumeFormat(pool) {
+
+/**
+ * Returns whetever disk property of VM's inactive XML has changed
+ * comparted to live XML.
```
typo compared.

KKoukiou commented on pkg/machines/helpers.js:
```diff
@@ -676,3 +676,77 @@ export function getDefaultVolumeFormat(pool) {
+}
+
+/**
+ * Returns a shorted pretty version of disk name.
```
shortened

KKoukiou commented on pkg/machines/libvirt-dbus.js:
```diff
@@ -1338,6 +1339,14 @@ export function attachIface({ connectionName, vmId, mac, permanent, hotplug, sou
     return attachDevice({ connectionName, vmId, permanent, hotplug, xmlDesc });
 }
 
+export function changeDiskSettings(connectionName, objPath, target, readonly, shareable) {
```
Maybe better changeDiskAccessPolicy is better, but I don't really mind this name either.

KKoukiou commented on pkg/machines/components/diskEdit.jsx:
```diff
@@ -0,0 +1,213 @@
+            <Modal id={`${idPrefix}-dialog`} show onHide={this.props.close}>
+                <Modal.Header>
+                    <Modal.CloseButton onClick={this.props.close} />
+                    <Modal.Title> {'Edit ' + getDiskPrettyName(vm.disks[disk.target]) + ' Attributes'} </Modal.Title>
```
I don't like the Idea of having 'Attributes' in the title, but if it was recommended by the designers I withdraw :)

skobyda commented:
Screenshots for release notes:

![Screenshot from 2019-10-02 13-16-31](https://user-images.githubusercontent.com/42733240/66040343-7e733a00-e517-11e9-8bba-e42ca313f21d.png)
![Screenshot from 2019-10-02 13-17-03](https://user-images.githubusercontent.com/42733240/66040344-7e733a00-e517-11e9-9216-9b8f31b3bec8.png)
![Screenshot from 2019-10-02 13-17-42](https://user-images.githubusercontent.com/42733240/66040345-7e733a00-e517-11e9-821e-7e4678ed42cb.png)
